### PR TITLE
fix(cis): retry get shadow operation

### PR DESCRIPTION
**Pull request by jcosentino11**
**Issue #, if available:**

**Description of changes:**

On startup, CIS shadow monitor requests the latest CIS shadow via an MQTT request, and will update certificates with host information accordingly.  However, there is a subtle race where nucleus' mqtt client may reset during this CIS shadow get operation (e.g. due to device configuration changing).

This PR adds retry logic when fetching CIS shadows, to ensure we are resilient to failures.  We can simply wait until we receive a get shadow response, which is either:
* get shadow accepted
* get shadow rejected

Accepted is the happy path. We requested a shadow and we received it.  Rejected means the shadow doesn't exist yet; there's no need to retry because we'll get a delta update when the shadow is eventually created.

**Why is this change necessary:**
If get shadow fails when CDA starts up, server certificates may not be populated with the proper SANs, causing TLS connections between clients and local broker to fail.

**How was this change tested:**

Unit tests

**Any additional information or context required to review the change:**

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java [hunk 1]
```diff
@@ -67,6 +71,21 @@ public class CISShadowMonitor implements Consumer<NetworkStateProvider.Connectio
...
+            RetryUtils.RetryConfig.builder()
+                    .initialRetryInterval(Duration.ofSeconds(1L))
+                    .maxRetryInterval(Duration.ofMinutes(30L))
+                    .maxAttempt(Integer.MAX_VALUE)
+                    .retryableExceptions(Collections.singletonList(Exception.class))
+                    .build();
+
+    @Getter(AccessLevel.PACKAGE)
+    @Setter(AccessLevel.PACKAGE)
+    private long mqttOperationTimeoutSeconds = 10L;
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:84]: get the timeout from the mqtt client. mqttClient.getOperationTimeout or something like that

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:84]: In fact, our MQTT client already has a timeout; so I'm not sure we actually need our own timeout here, it may be fine to just do a blocking `get`

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java [hunk 2]
```diff
@@ -158,21 +177,37 @@ private void subscribeToShadowTopics() throws InterruptedException {
                 ShadowDeltaUpdatedSubscriptionRequest shadowDeltaUpdatedSubscriptionRequest =
                         new ShadowDeltaUpdatedSubscriptionRequest();
                 shadowDeltaUpdatedSubscriptionRequest.thingName = shadowName;
-                iotShadowClient.SubscribeToShadowDeltaUpdatedEvents(shadowDeltaUpdatedSubscriptionRequest,
+                iotShadowClient.SubscribeToShadowDeltaUpdatedEvents(
+                                shadowDeltaUpdatedSubscriptionRequest,
                                 QualityOfService.AT_LEAST_ONCE,
                                 this::processCISShadow,
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:183]: store the callback as a final field to prevent recreating lambdas.

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java [hunk 3]
```diff
@@ -158,21 +177,37 @@ private void subscribeToShadowTopics() throws InterruptedException {
...
                 getShadowSubscriptionRequest.thingName = shadowName;
-                iotShadowClient.SubscribeToGetShadowAccepted(getShadowSubscriptionRequest,
-                                QualityOfService.AT_LEAST_ONCE, this::processCISShadow,
+                iotShadowClient.SubscribeToGetShadowAccepted(
+                                getShadowSubscriptionRequest,
+                                QualityOfService.AT_LEAST_ONCE,
+                                resp -> {
+                                    reportShadowReceived();
+                                    processCISShadow(resp);
+                                },
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:197]: store callback as a final field to prevent recreating lambdas which may lead to duplicate calls on message receipt

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java [hunk 4]
```diff
@@ -158,21 +177,37 @@ private void subscribeToShadowTopics() throws InterruptedException {
...
-                        .get(TIMEOUT_FOR_SUBSCRIBING_TO_TOPICS_SECONDS, TimeUnit.SECONDS);
+                        .get(mqttOperationTimeoutSeconds, TimeUnit.SECONDS);
                 LOGGER.info("Subscribed to shadow get accepted topic");
+
+                GetShadowSubscriptionRequest getShadowRejectedSubscriptionRequest = new GetShadowSubscriptionRequest();
+                getShadowRejectedSubscriptionRequest.thingName = shadowName;
+                iotShadowClient.SubscribeToGetShadowRejected(
+                                getShadowRejectedSubscriptionRequest,
+                                QualityOfService.AT_LEAST_ONCE,
+                                err -> reportShadowReceived(),
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:207]: same.

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java [hunk 5]
```diff
@@ -315,14 +351,60 @@ private void updateCISShadowReportedState(Map<String, Object> reportedState) {
...
+                // operation already in progress
+                return;
+            }
+            getShadowTask = executorService.submit(() -> {
+                try {
+                    RetryUtils.runWithRetry(
+                            GET_CIS_SHADOW_RETRY_CONFIG,
+                            () -> {
+                                CompletableFuture<?> shadowReceived = this.shadowGetResponseReceived.updateAndGet(
+                                        ignore -> new CompletableFuture<>());
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:374]: just do a set instead of update?

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java [hunk 6]
```diff
@@ -315,14 +351,60 @@ private void updateCISShadowReportedState(Map<String, Object> reportedState) {
...
+
+    private CompletableFuture<Integer> publishToGetCISShadowTopic() {
         LOGGER.atDebug().log("Publishing to get shadow topic");
         GetShadowRequest getShadowRequest = new GetShadowRequest();
         getShadowRequest.thingName = shadowName;
-        iotShadowClient.PublishGetShadow(getShadowRequest, QualityOfService.AT_LEAST_ONCE).exceptionally(e -> {
-            LOGGER.atWarn().cause(e).log("Unable to retrieve CIS shadow");
-            return null;
-        });
+        return iotShadowClient.PublishGetShadow(getShadowRequest, QualityOfService.AT_LEAST_ONCE)
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:403]: use qos 0. Otherwise when we're offline this is going to fill up the spooler.

you may also want to not retry when we know that we're offline.

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java [hunk 7]
```diff
@@ -315,14 +351,60 @@ private void updateCISShadowReportedState(Map<String, Object> reportedState) {
...
+        CompletableFuture<?> shadowReceived = this.shadowGetResponseReceived.get();
+        if (shadowReceived != null) {
+            shadowReceived.complete(null);
+        }
+    }
+
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
+    private void fetchCISShadowWithRetriesAsync() {
+        synchronized (getShadowLock) {
+            if (getShadowTask != null && !getShadowTask.isDone()) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:364]: check for cancelled? or does done include cancelled

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java (hunk 1 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:84]: ah yeah, will adjust

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java (hunk 6 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:403]: good point. and yeah when cda goes offline the task will be cancelled, but i didn't handle case if we start offline will fix

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java (hunk 7 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:364]: done includes cancelled

**Comment by jcosentino11** [comment]
comments addressed, still need to update the test, had to step out for a little bit

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java [hunk 8]
```diff
@@ -33,45 +33,70 @@
...
-    private static final long TIMEOUT_FOR_SUBSCRIBING_TO_TOPICS_SECONDS = Duration.ofMinutes(1).getSeconds();
-    private static final long WAIT_TIME_TO_SUBSCRIBE_AGAIN_IN_MS = Duration.ofMinutes(2).toMillis();
-    private static final Random JITTER = new Random();
     static final String SHADOW_UPDATE_DELTA_TOPIC = "$aws/things/%s/shadow/update/delta";
     static final String SHADOW_GET_ACCEPTED_TOPIC = "$aws/things/%s/shadow/get/accepted";
 
+    private static final RetryUtils.RetryConfig SUBSCRIBE_TO_TOPICS_RETRY_CONFIG =
+            RetryUtils.RetryConfig.builder()
+                    .initialRetryInterval(Duration.ofSeconds(1L))
+                    .maxRetryInterval(Duration.ofMinutes(30L))
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:66]: 30 minutes? Seems a bit much, no?

If we're only trying when we know that we're online, then retrying faster should be alright.

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java [hunk 9]
```diff
@@ -149,49 +208,38 @@ public void removeFromMonitor(CertificateGenerator certificateGenerator) {
...
-                LOGGER.atWarn().setCause(e).log("Subscribe to shadow topics timed out, will retry shortly");
-            }
-            // Wait for sometime and then try to subscribe again
-            Thread.sleep(WAIT_TIME_TO_SUBSCRIBE_AGAIN_IN_MS + JITTER.nextInt(10_000));
-        }
+    private void subscribeToShadowTopics() throws InterruptedException, ExecutionException {
+        ShadowDeltaUpdatedSubscriptionRequest shadowDeltaUpdatedSubscriptionRequest =
+                new ShadowDeltaUpdatedSubscriptionRequest();
+        shadowDeltaUpdatedSubscriptionRequest.thingName = shadowName;
+        iotShadowClient.SubscribeToShadowDeltaUpdatedEvents(
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:215]: let's do it in a different PR, but we should stop using the shadow client. It dynamically creates a new message handler every time it is called which will lead to duplicate subscription callbacks.

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java [hunk 10]
```diff
@@ -149,49 +208,38 @@ public void removeFromMonitor(CertificateGenerator certificateGenerator) {
...
+
+        GetShadowSubscriptionRequest getShadowSubscriptionRequest = new GetShadowSubscriptionRequest();
+        getShadowSubscriptionRequest.thingName = shadowName;
+        iotShadowClient.SubscribeToGetShadowAccepted(
+                        getShadowSubscriptionRequest,
+                        QualityOfService.AT_LEAST_ONCE,
+                        onGetShadowAccepted,
+                        (e) -> LOGGER.atError().log("Error processing getShadowSubscription Response", e))
+                .get();
+        LOGGER.atInfo().log("Subscribed to shadow get accepted topic");
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:232]: debug

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java (hunk 8 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:66]: in this case it's retrying even when offline, but could probably be smarter about this, will revise

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java (hunk 9 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:215]: ah geez, yep will follow up w/ a fix

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java (hunk 8 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java]: fixed. also consolidated subscribing future and get future into a single future (making subscribe only happen once).  also removed unsubscribing since it's unnecessary

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java [hunk 11]
```diff
@@ -33,45 +33,88 @@
...
+                    .maxRetryInterval(Duration.ofSeconds(30L))
+                    .maxAttempt(Integer.MAX_VALUE)
+                    .retryableExceptions(Collections.singletonList(Exception.class))
+                    .build();
+
+    // we only need to subscribe to shadow topics once across the lifetime of this monitor
+    private final AtomicBoolean subscribed = new AtomicBoolean();
+
+    // we don't need to unsubscribe, because mqtt client reconnects with clean session on startup.
+    // instead, we can cancel current operations and block new ones on shutdown.
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:86]: on reconnect though, the subscriptions which were successful will be re-subscribed by the mqtt client for you.

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java [hunk 12]
```diff
@@ -33,45 +33,88 @@
...
+    private final AtomicBoolean stopped = new AtomicBoolean();
+    private final Consumer<ShadowDeltaUpdatedEvent> onShadowDeltaUpdated = event -> {
+        if (stopped.get()) {
+            return;
+        }
+        processCISShadow(event);
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:92]: if we got an event, why would we ignore it?

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java [hunk 13]
```diff
@@ -33,45 +33,88 @@
...
+    };
+    private final Consumer<GetShadowResponse> onGetShadowAccepted = resp -> {
+        if (stopped.get()) {
+            return;
+        }
+        reportShadowReceived();
+        processCISShadow(resp);
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:99]: why would we ignore any response?

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java [hunk 14]
```diff
@@ -33,45 +33,88 @@
...
+    };
+    private final Consumer<ErrorResponse> onGetShadowRejected = err -> {
+        if (stopped.get()) {
+            return;
+        }
+        reportShadowReceived();
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:105]: report shadow received? But this is rejected...

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java (hunk 12 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:92]: this was to handle shutdown w/o unsubscribe

Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:92]: will add unsubscribe back in, looks to be less confusing---was originally thinking on the terms of not having extra mqtt calls if after shutdown we'd lose the subscription anyway (since nucleus first-time connect is w/ clean session)

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java (hunk 14 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:105]: bad naming, it's to signal that we've received any shadow response, whether accepted or rejected, will think of better name

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java (hunk 12 above)
Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:92]: If we're shutting down, why does it matter then if we process the event or not?

Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:92]: Was to avoid extra processing if we receive a shadow update after the task was already cancelled. 

I guess it's not a big deal though because nucleus will shutdown the executors eventually

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java (hunk 11 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java]: yep, this was talking about shutdown and startup.  but just to confirm my thinking, unsubscribe on shutdown is technically unnecessary right?

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java]: Yes, unsubscribing on shutdown shouldn't really matter.

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java [hunk 15]
```diff
@@ -81,54 +108,60 @@ public class CISShadowMonitor implements Consumer<NetworkStateProvider.Connectio
...
 
     /**
      * Stop shadow monitor.
      */
     public void stopMonitor() {
-        if (subscribeTaskFuture != null) {
-            subscribeTaskFuture.cancel(true);
-        }
-        unsubscribeFromShadowTopics();
+        subscribed.set(false);
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:163]: why set it to false? You aren't ever unsubscribing, so you are still subscribed.

Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java]: revised the subscription stuff

File: src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java [hunk 16]
```diff
@@ -33,102 +35,169 @@
...
+                    .maxRetryInterval(Duration.ofSeconds(30L))
+                    .maxAttempt(Integer.MAX_VALUE)
+                    .retryableExceptions(Collections.singletonList(Exception.class))
+                    .build();
     private static final RetryUtils.RetryConfig GET_CONNECTIVITY_RETRY_CONFIG =
             RetryUtils.RetryConfig.builder().initialRetryInterval(Duration.ofMinutes(1L))
                     .maxRetryInterval(Duration.ofMinutes(30L)).maxAttempt(Integer.MAX_VALUE)
                     .retryableExceptions(Arrays.asList(ThrottlingException.class, InternalServerException.class))
                     .build();
+    private static final RetryUtils.RetryConfig GET_CIS_SHADOW_RETRY_CONFIG =
```
Comment (sameerzuberi) [inline src/main/java/com/aws/greengrass/clientdevices/auth/connectivity/CISShadowMonitor.java:74]: nit: do we need to separate the retry configs here? Looks like `SUBSCRIBE_TO_TOPICS` and `GET_CONNECTIVITY` are the same, could be combined in a single variable

**Review by vabhasin — APPROVED** [review]

**Review by sameerzuberi — APPROVED** [review]

**Review by sameerzuberi — APPROVED** [review]

**Review by vabhasin — APPROVED** [review]
